dlio_benchmark/reader/dali_tfrecord_reader.py

```diff
@@ -0,0 +1,105 @@
+from dlio_profiler.logger import dlio_logger as PerfTrace, fn_interceptor as Profile
+
+dlp = Profile(MODULE_DATA_READER)
+
+
+class DaliTFRecordReader(FormatReader):
+    @dlp.log_init
+    def __init__(self, dataset_type, thread_index, epoch):
+        super().__init__(dataset_type, thread_index)
+    def open(self):
+        super().open()
+
+    def close(self):
+        super().close()
+    
+    def get_sample(self, filename, sample_index):
+        super().get_sample(filename, sample_index)
+
+    def next(self):
+        super().next()
+
+    def read_index(self):
+        super().read_index()
+    @dlp.log
+    def read(self):
+        folder = "valid"
+        if self.dataset_type == DatasetType.TRAIN:
+            folder = "train"
+        index_folder = f"{self._args.data_folder}/index/{folder}"
+        index_files = []
+        for file in self._file_list:
+            filename = os.path.basename(file)
+            index_files.append(f"{index_folder}/{filename}.idx")
+        logging.info(
+            f"{utcnow()} Reading {len(self.file_list)} files rank {self._args.my_rank}")
+        random_shuffle = False
+        seed = -1
+        if self._args.sample_shuffle is not Shuffle.OFF:
+            if self._args.sample_shuffle is not Shuffle.SEED:
+                seed = self._args.seed
+            random_shuffle = True
+        initial_fill = 1024
+        if self._args.shuffle_size > 0:
+            initial_fill = self._args.shuffle_size
+        prefetch_size = 1
+        if self._args.prefetch_size > 0:
+            prefetch_size = self._args.prefetch_size
+        dataset = fn.readers.tfrecord(path=self._file_list,
+                                      index_path=index_files,
+                                      features={
+                                          'image': tfrec.FixedLenFeature((), tfrec.string, ""),
+                                          'size': tfrec.FixedLenFeature([1], tfrec.int64, 0)
+                                      }, num_shards=self._args.comm_size,
+                                      prefetch_queue_depth=prefetch_size,
+                                      initial_fill=initial_fill,
+                                      random_shuffle=random_shuffle, seed=seed,
+                                      stick_to_shard=True, pad_last_batch=True, 
+                                      dont_use_mmap=self._args.dont_use_mmap)
+        dataset = self._preprocess(dataset["image"])
+        dataset = self._resize(dataset)
+        return dataset
+
+    @dlp.log
+    def _preprocess(self, dataset):
+        if self._args.preprocess_time != 0. or self._args.preprocess_time_stdev != 0.:
+            t = np.random.normal(self._args.preprocess_time, self._args.preprocess_time_stdev)
+            sleep(max(t, 0.0))
+        return dataset
+
+    @dlp.log
+    def _resize(self, dataset):
+        return fn.resize(dataset, size=[self._args.max_dimension, self._args.max_dimension])
+
```

Comment:
Based on iterative or index set the correct method and throw error on other.

Reply:
Same treatment as above